src/keys/utils.ts

```diff
@@ -41,18 +47,71 @@ async function generateKeyPair(bits: number): Promise<KeyPair> {
   return await generateKeyPair({ bits });
 }
 
+function generateMnemonic(words: 12 | 24 = 24): string {
+  switch (words) {
+    case 12:
+      return Bip39.generateMnemonic(128);
+    case 24:
+    default:
+      return Bip39.generateMnemonic(256);
+  }
+}
+
+function generateSeedFromMnemonic(mnemonic: string) {
+  if (!isValidMnemonic(mnemonic)) throw new keysErrors.ErrorMnemonicInvalid();
+  return Bip39.mnemonicToEntropy(mnemonic);
+}
+
+function isValidMnemonic(mnemonic: string): boolean {
+  if (mnemonic.split(' ').length !== 24) return false;
+  return Bip39.validateMnemonic(mnemonic);
+}
+
+async function verifyMnemonicMatchesNode(
+  keyPair: KeyPair,
+  keysPath: string,
+  fs?: FileSystem,
+): Promise<boolean> {
+  const fs_ = fs ?? require('fs');
+
+  // We don't have a KeyManager to get the NodeId from, so we need to generate it.
+  const nodeId = publicKeyToFingerprint(keyPair.publicKey);
+
+  // Getting the rootcert.
+  const rootCertPath = path.join(keysPath, 'root.crt');
+  let rootCertPem;
+  try {
+    rootCertPem = await fs_.promises.readFile(rootCertPath, {
+      encoding: 'utf8',
+    });
+  } catch (e) {
+    throw new keysErrors.ErrorRootCertRead(e.message, {
+      errno: e.errno,
+      syscall: e.syscall,
+      code: e.code,
+      path: e.path,
+    });
+  }
+  const rootCert = certFromPem(rootCertPem);
+  const litmusNodeId = networkUtils.certNodeId(rootCert);
+
+  // Checking if nodeIds match.
+  return nodeId === litmusNodeId;
+}
+
 async function generateDeterministicKeyPair(
   bits: number,
-  seed: string,
+  mnemonic: string,
 ): Promise<KeyPair> {
+  const seed = generateSeedFromMnemonic(mnemonic);
   const prng = random.createInstance();
   prng.seedFileSync = (needed: number) => {
     // Using bip39 seed generation parameters
     // no passphrase is considered here
     return pkcs5.pbkdf2(seed, 'mnemonic', 2048, needed, md.sha512.create());
   };
   const generateKeyPair = promisify(pki.rsa.generateKeyPair).bind(pki.rsa);
-  return await generateKeyPair({ bits, prng });
+  return await generateKeyPair({ bits, prng, workers: -1 });
```

Comment:
What does `workers: -1` do here? Does node-forge have its own worker threads?

Reply:
I noticed you were doing PBKDF2 twice here as `bip39.mnemonicToSeedSync` does this already. So we use the mnemonic directly and this is sufficient.

Reply:
New function:

```ts
async function generateDeterministicKeyPair(
  bits: number,
  recoveryCode: string,
): Promise<KeyPair> {
  const prng = random.createInstance();
  prng.seedFileSync = (needed: number) => {
    // Using bip39 seed generation parameters
    // no passphrase is considered here
    return pkcs5.pbkdf2(recoveryCode, 'mnemonic', 2048, needed, md.sha512.create());
  };
  const generateKeyPair = promisify(pki.rsa.generateKeyPair).bind(pki.rsa);
  return await generateKeyPair({ bits, prng });
}
```

Reply:
So at this point, you can remove anything that uses `mnemonicToSeedSync` or `mnemonicToEntropy` none of these need to be used.

The `validateRecoveryCode` can still be used. And `generateRecoveryCode` should still be used though.

Reply:
I've fixed it up according to https://github.com/MatrixAI/js-polykey/issues/202#issuecomment-878775248

Reply:
The `workers: -1` didn't seem to do anything different for us, and I'd hesitate to rely on worker threads provided by node-forge, don't trust it.